src/Features/Core/Shared/Extensions/DiagnosticAnalyzerExtensions.cs

```diff
@@ -54,7 +54,12 @@ public static DiagnosticAnalyzerCategory GetDiagnosticAnalyzerCategory(this Diag
                         category |= DiagnosticAnalyzerCategory.ProjectAnalysis;
                     }
 
-                    if (!cantSupportSemanticSpanAnalysis)
+                    // Current DiagnosticAnalyzer API has no restriction on any analyzer action reporting diagnostics anywhere within a compilation.
+                    // Hence, we must conservatively assume they can't support incremental span based analysis across compilations.
+                    // GitHub issue https://github.com/dotnet/roslyn/issues/1565 tracks adding support for marking an analyzer incremental.
+                    // Until then, we just assume that all public/non-built in analyzers can never support incremental analysis.
+
+                    if ((analyzer is IBuiltInAnalyzer) && !cantSupportSemanticSpanAnalysis)
```

Comment:
For example, I have a diagnostic that wants to make sure that a particular method is invoked only once in a type (other cases should rely on cached value from it).
Correct way to implement it is via a compilation end action, tracking state.
However, I can also write it as a symbol/syntax node analyzer for named types, that walks all descendant nodes, binds invocation expressions, and if more than 1 expressions bind to that method, fire a diagnostic on all but first occurrence. Our current heuristic will mark this analyzer as candidate for incremental method body analysis, but that's incorrect as any edits in method body can invalidate diagnostics reported on other method body in this type.